### PR TITLE
Add CSV support for address and name query params

### LINKED ISSUE #1438: Add CSV support for address and name query params
### Description

Extend the request query params support on the following endpoints:

```
HTTP GET /api/v3/pos/validators
RPC  get.pos.validators
```

### Acceptance Criteria

- Allow users to perform query validators with CSV inputs on the following request params
  - `address`
  - `name`
- Gateway pattern checks are up-to-date
  - Allow a maximum of 10 addresses in CSV (max 419 chars)
  - Allow a maximum of 10 names in CSV (max 209 chars)
- New integration tests are added
- Swagger (description) is up-to-date

### PR DESCRIPTION
**Pull request by priojeetpriyom**
### What was the problem?
This PR resolves #1438

### How was it solved?
- Allow users to perform query validators with CSV inputs on the following request params
    - [x] address
    - [x] name
- Gateway pattern checks are up-to-date
    - [x] Allow a maximum of 10 addresses in CSV (max 419 chars)
    - [x] Allow a maximum of 10 names in CSV (max 209 chars)
- [x] Add integration test for previous cases
- [x] New integration tests are added
- [x] Swagger (description) is up-to-date
- [x] Refactor hardcoded api param mappings
- [ ] Add `altSwaggerKey: 'validatorSearch'` to `search` param of `stakers` and `stakes` endpoint once #1446 is merged.

### How was it tested?
- [x] Local

### CODE REVIEW

File: services/blockchain-indexer/shared/dataService/pos/validators.js [hunk 1]
```diff
@@ -187,32 +187,46 @@ const getPosValidators = async params => {
...
-			? (acc[entity] && params[entity].split(',').includes(acc[entity]))
-			: (acc[entity] && acc[entity] === params[entity]),
-	);
+	const addressSet = new Set();
+	const nameSet = new Set();
 
 	if (params.publicKey) {
-		params.address = getLisk32AddressFromPublicKey(params.publicKey);
+		const address = getLisk32AddressFromPublicKey(params.publicKey);
+		params.address = params.address || address;
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/dataService/pos/validators.js:195]: `params.address` should not be available when the user queries with `params.publicKey`. Did we not define the param pairings in the gateway?
```suggestion
		params.address = address;
```

File: services/gateway/shared/utils.js [hunk 2]
```diff
@@ -23,6 +23,14 @@ const transformParams = (type, params) => {
...
+		if (params[paramKey].altSwaggerKey) {
+			data.push({
+				$ref: `#/parameters/${params[paramKey].altSwaggerKey}`,
+			});
+			return;
+		}
+
 		if (type === 'blocks' && paramKey === 'id') {
 			value = { $ref: '#/parameters/block' };
 		} else if (type === 'PoS' && paramKey === 'status') {
```
Comment (sameersubudhi) [inline services/gateway/shared/utils.js:36]: Could you please set the `altSwaggerKey` for these special cases and entirely remove the `if...else` ladder here?

File: services/blockchain-indexer/shared/dataService/pos/validators.js [hunk 3]
```diff
@@ -187,32 +187,46 @@ const getPosValidators = async params => {
...
+	const validatorsTable = await getValidatorsTable();
+	validators.data = await getAllValidators();
+
+	const allValidators = await getAllValidators();
+
+	// Filter validators based on user passed params
+	const filteredValidators = allValidators.filter(validator => {
+		if (addressSet.size && !addressSet.has(validator.address)) return false;
+		if (nameSet.size && !nameSet.has(validator.name)) return false;
+		if (params.status && params.status !== validator.status) return false;
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/dataService/pos/validators.js:221]: Please verify if `params.status` takes in a CSV input and handle the scenario accordingly.

File: services/gateway/apis/http-version3/methods/modules/pos/posValidators.js [hunk 4]
```diff
@@ -24,9 +24,9 @@ module.exports = {
 	rpcMethod: 'get.pos.validators',
 	tags: ['PoS'],
 	params: {
-		address: { optional: true, type: 'string', pattern: regex.ADDRESS_LISK32 },
+		address: { optional: true, type: 'string', pattern: regex.ADDRESS_LISK32_CSV, max: 419, altSwaggerKey: 'accountAddressCsv' },
```
Comment (sameersubudhi) [inline services/gateway/apis/http-version3/methods/modules/pos/posValidators.js:27]:
```suggestion
		address: { optional: true, type: 'string', pattern: regex.ADDRESS_LISK32_CSV, max: 419, altSwaggerKey: 'addressCsv' },
```

File: services/gateway/apis/http-version3/methods/modules/pos/posValidators.js [hunk 5]
```diff
@@ -24,9 +24,9 @@ module.exports = {
...
 		publicKey: { optional: true, type: 'string', pattern: regex.PUBLIC_KEY },
-		name: { optional: true, type: 'string', pattern: regex.NAME },
+		name: { optional: true, type: 'string', pattern: regex.NAME_CSV, max: 209, altSwaggerKey: 'validatorNameCsv' },
```
Comment (sameersubudhi) [inline services/gateway/apis/http-version3/methods/modules/pos/posValidators.js:29]:
```suggestion
		name: { optional: true, type: 'string', pattern: regex.NAME_CSV, max: 209, altSwaggerKey: 'nameCsv' },
```

File: services/gateway/apis/http-version3/swagger/parameters/common.json [hunk 6]
```diff
@@ -1,8 +1,25 @@
 {
+    "accountAddressCsv": {
```
Comment (sameersubudhi) [inline services/gateway/apis/http-version3/swagger/parameters/common.json:2]:
```suggestion
    "addressCsv": {
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: services/gateway/apis/http-version3/methods/modules/pos/posValidators.js (hunk 5 above)
Comment (priojeetpriyom) [reply services/gateway/apis/http-version3/methods/modules/pos/posValidators.js:29]: I would prefer keeping previous name as the description is specific to validators.

```
Query validators by their registered name (can be expressed as CSV)
```

File: services/blockchain-indexer/shared/dataService/pos/validators.js (hunk 3 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/pos/validators.js:221]: Applied.

File: services/gateway/apis/http-version3/methods/modules/pos/posValidators.js (hunk 4 above)
Comment (priojeetpriyom) [reply services/gateway/apis/http-version3/methods/modules/pos/posValidators.js:27]: Applied.

File: services/gateway/apis/http-version3/swagger/parameters/common.json (hunk 6 above)
Comment (priojeetpriyom) [reply services/gateway/apis/http-version3/swagger/parameters/common.json:2]: Applied.

File: services/gateway/apis/http-version3/methods/modules/pos/posValidators.js (hunk 5 above)
Comment (sameersubudhi) [reply services/gateway/apis/http-version3/methods/modules/pos/posValidators.js:29]: Please ensure that adding an additional property is not causing issues with the `fastest-validator` implementation.

File: services/blockchain-indexer/shared/dataService/pos/validators.js (hunk 1 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/pos/validators.js:196]: we do not have param pairings defined in gateway. When called without params, this api returns a list of 10 validators. Also, param pairings validate at least one of the allowed must be passed, but do not validate that only one can be passed.

**Comment by priojeetpriyom** [comment]
TODO: Add `altSwaggerKey: 'validatorSearch'` to `search` param of `stakers` and `stakes` endpoint once #1446 is merged.

File: services/gateway/shared/utils.js (hunk 2 above)
Comment (priojeetpriyom) [reply services/gateway/shared/utils.js]: Applied.

Comment (priojeetpriyom) [reply services/gateway/shared/utils.js]: Applied.

File: services/gateway/apis/http-version3/methods/modules/pos/posValidators.js (hunk 5 above)
Comment (priojeetpriyom) [reply services/gateway/apis/http-version3/methods/modules/pos/posValidators.js:29]: Working fine with current validation.

File: services/gateway/apis/http-version3/swagger/parameters/common.json [hunk 7]
```diff
@@ -1,8 +1,25 @@
 {
+    "addressCsv": {
+        "name": "address",
+        "in": "query",
+        "description": "Address of the account to query (can be expressed as CSV).",
+        "type": "string",
+        "format": "address",
+        "minLength": 41
```
Comment (nagdahimanshu) [inline services/gateway/apis/http-version3/swagger/parameters/common.json:8]: May be we can also add maxLength:
```suggestion
        "minLength": 41,
        "maxLength": 419
```

File: services/gateway/apis/http-version3/swagger/parameters/pos.json [hunk 8]
```diff
@@ -13,6 +13,28 @@
...
+		"description": "Query account by the registered name.",
+		"type": "string",
+		"minLength": 1
+	},
+	"validatorNameCsv": {
+		"name": "name",
+		"in": "query",
+		"description": "Query validators by their registered name (can be expressed as CSV).",
+		"type": "string",
+		"minLength": 1
```
Comment (nagdahimanshu) [inline services/gateway/apis/http-version3/swagger/parameters/pos.json:28]:
```suggestion
		"minLength": 1,
		 "maxLength": 209
```

File: tests/integration/api_v3/http/posValidators.test.js [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+	beforeAll(async () => {
+		const response = await api.get(`${baseUrlV3}/pos/constants`);
+		numberActiveValidators = response.data.numberActiveValidators;
+		numberStandbyValidators = response.data.numberStandbyValidators;
+
+		const generatorsResponse = await api.get(`${baseUrlV3}/generators?limit=10`);
+		refGenerators = generatorsResponse.data;
+	});
+
+	describe(`GET ${endpoint}`, () => {
```
Comment (nagdahimanshu) [inline tests/integration/api_v3/http/posValidators.test.js:41]: Don't we already have test for validators?

File: services/blockchain-indexer/shared/dataService/pos/validators.js (hunk 1 above)
Comment (sameersubudhi) [reply services/blockchain-indexer/shared/dataService/pos/validators.js:196]: For now, please add a TODO to verify the param pairings logic and implement the same for the endpoint. Also, set `params.address = address;` here.

File: services/gateway/apis/http-version3/swagger/parameters/pos.json [hunk 10]
```diff
@@ -13,6 +13,28 @@
...
+	},
+	"validatorSearch": {
+		"name": "search",
+		"in": "query",
+		"description": "Validator name full text search phrase.",
```
Comment (sameersubudhi) [inline services/gateway/apis/http-version3/swagger/parameters/pos.json:33]:
```suggestion
		"description": "Validator name text search phrase. Supports both partial and full text search.",
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: tests/integration/api_v3/http/posValidators.test.js (hunk 9 above)
Comment (priojeetpriyom) [reply tests/integration/api_v3/http/posValidators.test.js:41]: I could not find one. Hence added newly.

File: services/gateway/apis/http-version3/swagger/parameters/common.json (hunk 7 above)
Comment (priojeetpriyom) [reply services/gateway/apis/http-version3/swagger/parameters/common.json:8]: This condition may not be true for other places where we may use this description in future. will skip for this reason.

File: services/blockchain-indexer/shared/dataService/pos/validators.js (hunk 1 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/pos/validators.js]: Applied.

File: tests/schemas/api_v3/posValidators.schema.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+const validatorSchema = {
+	name: Joi.string().pattern(regex.NAME).required(),
+	totalStakeReceived: Joi.string().pattern(regex.POSITIVE_DIGITS).required(),
+	selfStake: Joi.string().pattern(regex.POSITIVE_DIGITS).required(),
+	validatorWeight: Joi.string().pattern(regex.POSITIVE_DIGITS).required(),
+	address: Joi.string().pattern(regex.ADDRESS_LISK32).required(),
+	lastGeneratedHeight: Joi.number().integer().min(0).required(),
+	status: Joi.string().pattern(regex.POS_VALIDATOR_STATUS).required(),
+	isBanned: Joi.boolean().required(),
+	pomHeights: Joi.array().items(Joi.number().integer().min(0)).required(),
```
Comment (nagdahimanshu) [inline tests/schemas/api_v3/posValidators.schema.js:38]: With alpha.12, `pomHeights ` is renamed to `reportMisbehaviorHeights `

File: tests/schemas/api_v3/posValidators.schema.js [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+};
+
+const sharingCoefficient = {
+	tokenID: Joi.string().pattern(regex.TOKEN_ID).required(),
+	coefficient: Joi.string().required(),
+};
+
+const validatorSchema = {
+	name: Joi.string().pattern(regex.NAME).required(),
+	totalStakeReceived: Joi.string().pattern(regex.POSITIVE_DIGITS).required(),
```
Comment (nagdahimanshu) [inline tests/schemas/api_v3/posValidators.schema.js:31]: `totalStakeReceived` is renamed to `totalStake`
```suggestion
	totalStake: Joi.string().pattern(regex.POSITIVE_DIGITS).required(),
```

File: tests/schemas/api_v3/posValidators.schema.js (hunk 11 above)
Comment (priojeetpriyom) [reply tests/schemas/api_v3/posValidators.schema.js]: Applied.

File: tests/schemas/api_v3/posValidators.schema.js (hunk 12 above)
Comment (priojeetpriyom) [reply tests/schemas/api_v3/posValidators.schema.js]: Applied.

**Review by nagdahimanshu — APPROVED** [review]
Looks good 👍

**Review by sameersubudhi — APPROVED** [review]

**Comment by sameersubudhi** [comment]
@priojeetpriyom I'm merging this PR, please address the following comment in PR https://github.com/LiskHQ/lisk-service/pull/1446 after you update the changes from the base branch.

> Add altSwaggerKey: 'validatorSearch' to search param of stakers and stakes endpoint once https://github.com/LiskHQ/lisk-service/pull/1446 is merged.

**Pull request merged**